algorithms/src/snark/marlin/ahp/prover/round_functions/second.rs

```diff
@@ -150,94 +215,104 @@ impl<F: PrimeField, MM: MarlinMode> AHPForR1CS<F, MM> {
             fft::Evaluations::from_vec_and_domain(r_alpha_x_evals, mul_domain)
         };
         multiplier.add_evaluation(r_alpha_x_evals, "r_alpha_x");
-        let mut lhs = multiplier
+        let lhs = multiplier
             .element_wise_arithmetic_4_over_domain(mul_domain, ["r_alpha_x", "summed_z_m", "z", "t"], |a, b, c, d| {
                 a * b - c * d
             })
             .unwrap();
 
-        lhs += &mask_poly.map_or(SparsePolynomial::zero(), |p| p.polynomial().as_sparse().unwrap().clone());
         end_timer!(q_1_time);
         lhs
     }
 
-    fn calculate_summed_z_m_and_t(
-        state: &prover::State<F, MM>,
+    fn calculate_summed_z_m_and_t<'a>(
+        state: &prover::State<'a, F, MM>,
         alpha: F,
         eta_b: F,
         eta_c: F,
-        batch_combiners: &[F],
-    ) -> (DensePolynomial<F>, DensePolynomial<F>) {
-        let constraint_domain = state.constraint_domain;
-        let summed_z_m_poly_time = start_timer!(|| "Compute z_m poly");
-
-        let fft_precomputation = &state.index.fft_precomputation;
-        let ifft_precomputation = &state.index.ifft_precomputation;
+        batch_combiners: &BTreeMap<&'a CircuitId, verifier::BatchCombiners<F>>,
```

Comment:
`CircuitId` can be `Copy`, so we don't need to pass it by reference

Reply:
Done